File: TS/TDD-Legacy-Code/E3.AddFuntionality/S3.Code/Calculator.ts

export enum OperationType {
    Plus,
    Div,
    Minus
}

export class Calculator {
    display: string = "";
    lastArgument = 0;
    result = 0;
    newArgument = false;
    shouldReset = true;

    lastOperation: OperationType;

    press(key: string) {
        if (key == "C") {
            this.display = "";
            return;
        }
        switch (key) {
            case "+":
                this.handleOpKey(OperationType.Plus);
                break;
            case "/":
                this.handleOpKey(OperationType.Div);
                break;
            case "-":
                this.handleOpKey(OperationType.Minus);
                break;
            case "=":
                let currentArgument = this.parseKeyNumber();
                if (this.lastOperation == OperationType.Plus) {
                    this.display = (this.lastArgument + currentArgument).toString();
                }
                if (this.lastOperation == OperationType.Div && currentArgument == 0) {
                    this.display = "Division By Zero Error";
                }
                this.shouldReset = true;
                break;
            default:
                if (this.shouldReset) {
                    this.display = "";
                    this.shouldReset = false;
                }
                if (this.newArgument) {
                    this.display = "";
                    this.newArgument = false;
                }
                this.display += key;
                break;
        }
    }

    private handleOpKey(op: OperationType) {
        this.lastOperation = op;
        this.lastArgument = this.parseKeyNumber();
        this.newArgument = true;
    }

    parseKeyNumber(): number {
        if (this.display == "")
            return 0;
        return parseInt(this.display);
    }

    getDisplay(): string {
        if (this.display == "")
            return "0";
        return this.display;
    }
}

File: Python/e3.addfunctionality/s1.refactor/s0/calculator.py
from enum import Enum


class OperationType(Enum):
    Plus = 1
    Div = 2


class Calculator:
    display = ""
    lastArgument = 0
    result = 0
    newArgument = False
    shouldReset = True
    lastOperation = OperationType.Plus

    def press(self, key):
        if key == "C":
            self.display = ""
        else:
            if key == "+":
                self.lastOperation = OperationType.Plus
                self.lastArgument = self.parseKeyNumber()
                self.newArgument = True
            else:
                if key == "/":
                    self.lastOperation = OperationType.Div
                    self.lastArgument = self.parseKeyNumber()
                    self.newArgument = True
                else:
                    if key == "=":
                        currentArgument = self.parseKeyNumber()
                        if self.lastOperation == OperationType.Plus:
                            self.display = str(self.lastArgument + currentArgument)
                        if self.lastOperation == OperationType.Div and currentArgument == 0:
                            self.display = "Division By Zero Error"
                        self.shouldReset = True
                    else:
                        if self.shouldReset:
                            self.display = ""
                            self.shouldReset = False
                        if self.newArgument:
                            self.display = ""
                            self.newArgument = False
                        self.display += key

    def get_display(self):
        if self.display == "":
            return "0"
        return self.display

    def parseKeyNumber(self):
        if self.display == "":
            return 0
        return int(self.display)
